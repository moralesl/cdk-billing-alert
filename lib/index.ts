import { Construct } from "@aws-cdk/core";
import * as cloudwatch from "@aws-cdk/aws-cloudwatch";
import * as cloudwatch_actions from "@aws-cdk/aws-cloudwatch-actions";
import * as sns from "@aws-cdk/aws-sns";
import * as subs from "@aws-cdk/aws-sns-subscriptions";

import { Duration } from "@aws-cdk/core";
import { ComparisonOperator } from "@aws-cdk/aws-cloudwatch";

export interface BillingAlertProps {
  /**
   * The emails that subscribe to the billing alert.
   */
  readonly emails: string[];

  /**
   * The amount of USD when the alarm should trigger.
   */
  readonly amount: number;

}

export class BillingAlert extends Construct {

  constructor(scope: Construct, id: string, props: BillingAlertProps) {
    super(scope, id);

    let billingMetric = new cloudwatch.Metric({
      metricName: "EstimatedCharges",
      namespace: "AWS/Billing",
      period: Duration.hours(6),
      statistic: "Maximum",
      dimensions: {
        Currency: "USD",
      },
    });

    const billingAlarm = billingMetric.createAlarm(scope, `BillingAlarm${props.amount}Dollars`, {
        alarmName: `Billing Alert - Estimated Bill Exceeds $${props.amount}`,
        alarmDescription: `Account Billing Alert for $${props.amount}`,
        threshold: props.amount,
        evaluationPeriods: 1,
        comparisonOperator:
          ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      }
    );

    const billingAlarmTopic = new sns.Topic(scope, `BillingAlert${props.amount}DollarsTopic`,
      {
        topicName: `Billing-Alert-${props.amount}Dollars`,
      }
    );

    // Add each mail to the subscription list
    props.emails.forEach((email) =>
      billingAlarmTopic.addSubscription(new subs.EmailSubscription(email))
    );

    billingAlarm.addAlarmAction(new cloudwatch_actions.SnsAction(billingAlarmTopic));
  }
}
