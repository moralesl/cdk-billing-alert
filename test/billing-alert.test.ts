import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { Stack } from '@aws-cdk/core';
import { BillingAlert } from '../lib/index';


test('BillingAlert matches snapshot', () => {
  const stack = new Stack();
  new BillingAlert(stack, 'TestBillingAlert', {
    amount: 100,
    emails: ["test@example.com"]
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('BillingAlert has mail subscription', () => {
  const stack = new Stack();

  new BillingAlert(stack, 'TestBillingAlert', {
    amount: 100,
    emails: ["test@example.com"]
  });

  expect(stack).toHaveResource('AWS::SNS::Subscription', {
    Endpoint: 'test@example.com'
  });
});

test('BillingAlert has amount as threshold', () => {
  const stack = new Stack();

  new BillingAlert(stack, 'TestBillingAlert', {
    amount: 42,
    emails: ["test@example.com"]
  });

  expect(stack).toHaveResource('AWS::CloudWatch::Alarm', {
    Threshold: 42
  });
});