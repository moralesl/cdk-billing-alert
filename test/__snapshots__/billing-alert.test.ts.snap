// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BillingAlert matches snapshot 1`] = `
Object {
  "Resources": Object {
    "BillingAlarm100Dollars0B7E2D29": Object {
      "Properties": Object {
        "AlarmActions": Array [
          Object {
            "Ref": "BillingAlert100DollarsTopic5FE6B983",
          },
        ],
        "AlarmDescription": "Account Billing Alert for $100",
        "AlarmName": "Billing Alert - Estimated Bill Exceeds $100",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "Currency",
            "Value": "USD",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "EstimatedCharges",
        "Namespace": "AWS/Billing",
        "Period": 21600,
        "Statistic": "Maximum",
        "Threshold": 100,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "BillingAlert100DollarsTopic5FE6B983": Object {
      "Properties": Object {
        "TopicName": "Billing-Alert-100Dollars",
      },
      "Type": "AWS::SNS::Topic",
    },
    "BillingAlert100DollarsTopictestexamplecom7976F141": Object {
      "Properties": Object {
        "Endpoint": "test@example.com",
        "Protocol": "email",
        "TopicArn": Object {
          "Ref": "BillingAlert100DollarsTopic5FE6B983",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
  },
}
`;
